### ISSUE TITLE
September Release nuget missing dependencies?

### ISSUE THREAD

**Issue by ThreadPitt** [SharePoint/PnP-Sites-Core#1330]
#### Category
[x ] Bug
[ ] Enhancement

#### Environment
[ ] Office 365 / SharePoint Online
[ x] SharePoint 2016
[ ] SharePoint 2013

If SharePoint on-premises, what's exact CU version: 
August 2017 CU (sts2016-kb4011049-fullfile-x64-glb)

I just updated the SharePointPnPCore2016 to the September release 2.18.1709.

When running the same code (a Visual Studio Console Application) as with the August release, I get the following error when calling ApplyProvisioningTemplate():

{"Could not load file or assembly 'Microsoft.IdentityModel.Clients.ActiveDirectory, Version=2.18.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies.
 The system cannot find the file specified.":"Microsoft.IdentityModel.Clients.ActiveDirectory, Version=2.18.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"}

When I installed that nuget package, I get:

{"Could not load file or assembly 'Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.":"Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"}

Then I gave up :-(

Cheers!

**Comment by jansenbe** [comment]
Good catch...we changed this for online but the change also is needed for the on-premises packages. I'll work on that asap

**Comment by jansenbe** [comment]
Just pushed out new NuGet packages for 2013/2016 PnP Sites Core version. The reason of this problem is the fact that the new serialization model depends on reflection which is very picky about having all the dependent assemblies present. I'm closing the issue, but feel free to open again if you still see this with version 2.18.1709.1

**Comment by myworkidentity** [comment]
I am also facing similar issues with version 2.20.1711.

Issue 1 - When calling **GetTemplate**, there is an exception. Stack trace: 

`at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XmlPnPSchemaBaseSerializer'1.DeserializeTemplate(Object persistenceTemplate, ProvisioningTemplate template)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XMLPnPSchemaV201705Serializer.DeserializeTemplate(Object persistenceTemplate, ProvisioningTemplate template)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XmlPnPSchemaBaseSerializer'1.ToProvisioningTemplate(Stream template, String identifier)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XMLPnPSchemaFormatter.ToProvisioningTemplate(Stream template, String identifier)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XMLTemplateProvider.GetTemplate(String uri, String identifier, ITemplateFormatter formatter, ITemplateProviderExtension[] extensions)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XMLTemplateProvider.GetTemplate(String uri, ITemplateProviderExtension[] extensions)
   at OfficeDevPnP.Core.Framework.Provisioning.Providers.Xml.XMLTemplateProvider.GetTemplate(String uri) `

Exception details reveal the following **LoaderExceptions** (the same exception is listed three times):

`{"Could not load file or assembly 'Microsoft.IdentityModel.Clients.ActiveDirectory, Version=2.29.0.1078, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.":"Microsoft.IdentityModel.Clients.ActiveDirectory, Version=2.29.0.1078, Culture=neutral, PublicKeyToken=31bf3856ad364e35"}`

Issue 2 - I also experienced a problem with Microsoft.Graph.Core reference. This exception was thrown when I was calling **UnifiedGroupsUtility.ListUnifiedGroups**:

`{"Could not load file or assembly 'Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.":"Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"}`

There was the following **FusionLog** entry listed for this:

`Pre-bind state information:
LOG: DisplayName = Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
 (Fully-specified)
LOG: Appbase = file:///G:/project/folder/Project.Project.SiteProvisioning/Project.Project.SiteProvisioning.Implementation/bin/Debug/
LOG: Initial PrivatePath = NULL
Calling assembly : OfficeDevPnP.Core, Version=2.20.1711.0, Culture=neutral, PublicKeyToken=3751622786b357c2.
LOG: This bind starts in default load context.
LOG: Using application configuration file: G:\project\folder\Project.Project.SiteProvisioning\Project.Project.SiteProvisioning.Implementation\bin\Debug\Project.Project.SiteProvisioning.Implementation.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework\v4.0.30319\config\machine.config.
LOG: Post-policy reference: Microsoft.Graph.Core, Version=1.2.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
LOG: Attempting download of new URL file:///G:/project/folder/Project.Project.SiteProvisioning/Project.Project.SiteProvisioning.Implementation/bin/Debug/Microsoft.Graph.Core.DLL.
LOG: Attempting download of new URL file:///G:/project/folder/Project.Project.SiteProvisioning/Project.Project.SiteProvisioning.Implementation/bin/Debug/Microsoft.Graph.Core/Microsoft.Graph.Core.DLL.
LOG: Attempting download of new URL file:///G:/project/folder/Project.Project.SiteProvisioning/Project.Project.SiteProvisioning.Implementation/bin/Debug/Microsoft.Graph.Core.EXE.
LOG: Attempting download of new URL file:///G:/project/folder/Project.Project.SiteProvisioning/Project.Project.SiteProvisioning.Implementation/bin/Debug/Microsoft.Graph.Core/Microsoft.Graph.Core.EXE.`

This issue 2 was "easily" averted by copying the current source code of **ListUnifiedGroups** and **CreateGraphClient** methods, and pasting them into my own solution. Of course, this is by far not an ideal solution, but I was able to solve the Microsoft.Graph.Core reference exception like this for the time being. 

I would hate to try solving also Issue 1 by copy-pasting the PnP source code into my own solution. I'd end up with a bunch of methods with names like **ListUnifiedGroups_CopyFromPnP**, **CreateGraphClient_CopyFromPnP**, and **GetTemplate_CopyFromPnP** and it would soon be a code maintainer's nightmare.

FYI - I have the following Nuget packages installed into my project: AngleSharp, AppForSharePointOnlineWebToolkit, AppForSharePointWebToolkit, Microsoft.AspNet.WebApi.Client, Microsoft.AspNet.WebApi.Core, Microsoft.Azure.ActiveDirectory.GraphClient, Microsoft.Azure.KeyVault.Core, Microsoft.Data.Edm, Microsoft.Data.OData, Microsoft.Data.Services.Client, Microsoft.Graph, Microsoft.Graph.Core, Microsoft.Identity.Client, Microsoft.IdentityModel, Microsoft.IdentityModel.Clients.ActiveDirectory, Microsoft.SharePointOnline.CSOM, MSFTCompressionCab, Newtonsoft.Json, NLog, SharePointPnP.IdentityModel.Extensions, SharePointPnPCoreOnline, System.Net.Http, System.Spatial, WindowsAzure.Storage. Each package with latest stable version, except Microsoft.Identity.Client which only has preview available. Maybe there is a conflict between some of these, or maybe not. 

Cheers,
myworkidentity

**Comment by myworkidentity** [comment]
No issues after all.... I was able to get the solution working by uninstalling **all** Nuget packages, then adding them back specifically starting with **SharePointPnPCoreOnline** package, followed by a few other Nugets that I was using. 

The whole solution then correctly references all the required versions of dependend packages. 

It is a little bit misleading that SharePointPnPCoreOnline Nuget package lists dependencies as e.g. "1.1.0.0=>" even if it is not entirely true - it absolutely seems to require a specific version, and updating latest Nuget versions of dependency packages **will not work**.

I also added a readme.txt to my solution to warn anyone maintaining the code to **not to update** any of the Nuget packages. Updating any dependent Nuget packages to latest versions will potentially break SharePointPnPCoreOnline package. Correct procedure is to uninstall everything, and then reinstall SharePointPnPCoreOnline and **to not touch any package updates** after that.
